0264-promise_gas_ratio.md

```diff
@@ -0,0 +1,145 @@
+
+Pros:
+- Can under-utilize the gas for the current transaction to limit gas allowed for certain functions 
+- This could take responsibility away from DApp users because they would not have to worry less about attaching too much prepaid gas
+- Thinking in terms of fractions may be more intuitive for some developers
+- Might future proof better if we ever need this ability in the future, want to minimize the number of host functions created at all costs
+
+Cons:
+- More complicated logic/edge cases to handle to make sure the percentages don't sum to greater than 100% (or adjusting if they do)
+- Precision loss from dividing integers may lead to unexpected results
+  - To get closer to expected, we could use floats for the division, but this gets messy
+- API for specifying a fraction would be messy (need to specify two values rather than just optionally one)
+- There isn't a good default for this. Unless there is a special value that indicates a pool of function calls that will split the remaining equally, but this defeats the purpose of this alternative completely
+- Slightly larger API (only one u64, can probably safely ignore this point)
+
+Alternative 2 (handle within contract/SDK):
+The other alternative is to handle all of this logic on the contract side, as seen by [this PR](https://github.com/near/near-sdk-rs/pull/523). This is much less feasible/accurate because there is only so much information available within the runtime, and gas costs and internal functionality may not always be the same. As discussed on [the respective issue](https://github.com/near/near-sdk-rs/issues/526), this alternative seems to be very infeasible.
+
+Pros:
+- No protocol change is needed
+- Can still have improved API as with protocol change
+
+Cons:
+- Additional bloat to every contract, even ones that don't use the pattern (~5kb in PoC, even with simple estimation logic)
+- Still inaccurate gas estimations, because at the point of calculation, we cannot know how much gas will be used for assigning gas values as well as gas consumed after the transaction ends
+  - This leads to either underutilizing or having transactions fail when using too much gas if trying to estimate how much gas will be left
+- Prone to breaking existing contracts on protocol changes that affect gas usage or logic of runtime
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+What needs to be addressed before this gets merged: 
+- How much refactoring exactly is needed to handle this pattern?
+    - Can we keep a queue of receipt and action indices with their respective ratios and update their gas values after the current method is executed? Is there a cleaner way to handle this while keeping order?
+- Do we want to attach the gas lost due to precision on division to any function?
```

Comment:
I'd say we just burn it. This is such a small amount of gas (for reasonable values of ratios) that doing anything else would just consume more gas than the amount that is being saved